app/controllers/flowchart_controller.rb

```diff
@@ -1,4 +1,103 @@
+  def create
+    @flowchart = Flowchart.new(JSON.parse(request.body.read))
+
+    unless @flowchart.valid?
+      render status: 400, json: { error: 'Invalid Flowchart params' }
+      return
+    end
+
+    @flowchart = Flowchart.create(JSON.parse(request.body.read))
+    root_node = FlowchartNode.create(flowchart_id: @flowchart.id, text: 'New Node', header: 'Options', is_root: true, deleted: false)
+    @flowchart[:root_id] = root_node.id
+    @flowchart.save
+
+    render json: @flowchart
+  end
+
+  def update
+    flowchart = Flowchart.find(params[:id])
+    if flowchart.nil?
+      render status: 404, json: { error: 'Could not find flowchart' }
+    end
+
+    flowchart.update_attributes(JSON.parse(request.body.read))
+    unless flowchart.valid?
+      render status: 400, json: { error: 'Invalid flowchart params' }
+      return
+    end
+
+    @flowchart = Flowchart.update(params[:id], JSON.parse(request.body.read))
+    render json: @flowchart
+  end
+
+  def delete
+    @flowchart = Flowchart.find(params[:id])
+    if @flowchart.nil?
+      render status: 404, json: { error: 'Could not find flowchart' }
+    end
+
+    Flowchart.find(params[:id]).update(deleted: true)
+    FlowchartNode.where(flowchart_id: params[:id]).update_all(deleted: true)
+    @flowchart[:deleted] = true
+
+    render json: @flowchart
+  end
+
+  def all_flowcharts
+    render json: Flowchart.where(deleted: false)
+  end
+
+  def serialized_flowchart_by_id
+    flowchartnodes = FlowchartNode.where(flowchart_id: params[:id], deleted: false)
+    flowchart = Flowchart.find_by(id: params[:id], deleted: false)
+    root_node = FlowchartNode.get_root_node(params[:id])
+
+    unless flowchart
+      render status: 404, json: { error: "No flowchart found with id #{params[:id]}." }
+      return
+    end
+
+    nodes_indexed_by_id = {}
+    flowchartnodes.each do |node|
+      nodes_indexed_by_id[node.id] = node
+    end
+
+    adjacency_list = {}
+    if root_node
+
+      queue = [root_node.id]
+      visited = {}
+      until queue.empty?
+        current_node_id = queue.shift
+        next if visited.key?(current_node_id)
+
+        visited[current_node_id] = true
+        current_node = nodes_indexed_by_id[current_node_id]
+
+        unless adjacency_list.key?(current_node_id)
+          adjacency_list[current_node_id] = {}
+        end
+
+        if current_node[:sibling_id]
+          adjacency_list[current_node_id][:sibling_id] = current_node[:sibling_id]
+          queue.push(current_node[:sibling_id])
+        end
+
+        if current_node[:child_id]
+          adjacency_list[current_node_id][:child_id] = current_node[:child_id]
+          queue.push(current_node[:child_id])
+        end
+      end
+
+    end
+
+    @serialized_flowchart = {}
+    @serialized_flowchart[:flowchart] = flowchart
+    @serialized_flowchart[:flowchartnodes] = nodes_indexed_by_id
+    @serialized_flowchart[:adjacency_list] = adjacency_list
```

Comment:
Why do we need to calculate `adjacency_list` here? It seems like the information it provides is already in `nodes_indexed_by_id`.

Reply:
Yeah I think it is better. The adjacency list lets you query the entire next level of node ids in the right order by just passing in the current node id. 

With nodes_indexed_by_id, the front end would either have to traverse the graph using sibling_id and child_id every time an option is selected (I think this is messy 
 because the code would probably have to go inside a react component) or the front end would to serialized the chart into an adjacency list (which is some logic that might as well live in the API).

Reply:
I might be misunderstanding how your adjacency list is being built, but it doesn't seem like you can query the entire next level of node ids in the right order from your adjacency list. In your `.get serialized` test, your expected output for the adjacency list is
```
'adjacency_list': {
          '1': {
            'child_id': 9
          },
          '3': {},
          '4': {
            'sibling_id': 3
          },
          '5': {},
          '6': {
            'child_id': 4
          },
          '8': {
            'child_id': 5
          },
          '9': {
            'sibling_id': 8,
            'child_id': 6
          }
        }
```
But, based on the way the graph is constructed, it seems like nodes `8`, `9`, `10` are all on the same level, but they don't appear as a level in the adjacency list.